restler/utils/logging/ndjson_serializer.py

```diff
@@ -0,0 +1,183 @@
+
+
+class CustomRotatingFileHandler(RotatingFileHandler):
+    def rotation_filename(self, default_name):
+        """Modify the rotation file name so the extension matches the default extension.
+        Note: this function must be idempotent.
+        """
+        base_name, count_ext = os.path.splitext(default_name)
+        base_name, type_ext = os.path.splitext(base_name)
+        return base_name + count_ext + type_ext
+
+class RequestTraceLog(RequestTraceLogBase):
+    def __init__(self, request_id=None, sequence_id=None, combination_id=None, tags={}):
+        self.request_id = request_id
+        self.sequence_id = sequence_id
+        self.combination_id = combination_id
+        self.sent_timestamp = None
+        self.received_timestamp = None
+        self.tags = tags
+        self._request = None
+        self._response = None
+        self.request_json = None
+        self.response_json = None
+
+    @classmethod
+    def from_dict(cls, log_dict):
+        instance = cls()
+        instance.request_id = log_dict.get('request_id')
+        instance.sequence_id = log_dict.get('sequence_id')
+        instance.combination_id = log_dict.get('combination_id')
+        instance.sent_timestamp = log_dict.get('sent_timestamp')
+        instance.received_timestamp = log_dict.get('received_timestamp')
+        instance.tags = log_dict.get('tags', {})
+        instance.request = log_dict.get('request')
+        instance.response = log_dict.get('response')
+        instance.request_json = json.loads(log_dict.get('request_json', '{}'))
+        instance.response_json = json.loads(log_dict.get('response_json', '{}'))
+        return instance
+
+    def normalize(self):
+        """ Normalize the request and response to a format that can be used to compare
+            request sequences across two different traces.
+            TODO: this is a placeholder implementation.  """
+        self.sent_timestamp = None
+        self.received_timestamp = None
+        self.sequence_id = None
+        self.tags = {}
+        return self
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            valueX = self.to_dict()
+            valueY = other.to_dict()
+            return valueX == valueY
+        else:
+            return False
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    @property
+    def request(self):
+        return self._request
+
+    @request.setter
+    def request(self, value):
+        self._request = value
+
+    @property
+    def response(self):
+        return self._response
+
+    @response.setter
+    def response(self, value):
+        self._response = value
+
+    def to_dict(self):
+        tags = {}
+        if self.request_id:
+            tags["request_id"] = self.request_id
+        if self.sequence_id:
+            tags["sequence_id"] = self.sequence_id
+        if self.combination_id:
+            tags["combination_id"] = self.combination_id
```

Comment:
I'm not sure if this is the right place for this comment, but would it be possible to include the origin too (e.g., "main_driver" or the checker's name)?

Apologies for not posting this comment sooner, I thought I had added it last week but I couldn't find it this morning.

Reply:
Yep!  Already planning to do this for the next iteration.  Let me know if you think of any others that would be useful.

Reply:
> Let me know if you think of any others that would be useful.

@marina-p, related to this request, while testing I found that the garbage collection thread is logging its requests and responses into the trace database. I tracked this down while investigating why my trace database had so many DELETE request/response pairs logged without `sequence_id`. I have `run_gc_after_every_sequence` set to true, which makes this easy to repro.

Assuming that the plan is to keep the GC thread activity included in the trace database, it would be helpful if there were an easy way to identify that these lines are coming from the GC thread. Is this something that could be included with the change to log the origin (maybe have a specific origin for the garbage collector)?

Thanks!

Reply:
Yep, will include this with the origin change.